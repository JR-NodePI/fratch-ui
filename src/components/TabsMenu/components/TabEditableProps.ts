export type TabEditableProps = {
  className?: string;
  editable?: boolean;
  label?: string;
  onActivate?: () => void;
  onChange?: (value: string) => void;
};
