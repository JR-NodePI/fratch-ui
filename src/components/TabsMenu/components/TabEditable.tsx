import { type FocusEvent, type KeyboardEvent, useState } from 'react';

import { c } from '../../../helpers';
import { IconEdit } from '../../Icons';
import { TabEditableProps } from './TabEditableProps';

import styles from './TabEditable.module.css';

export default function TabEditable({
  className,
  editable,
  label,
  onChange,
}: TabEditableProps): JSX.Element {
  const [isEditing, setIsEditing] = useState<boolean>(false);
  const [width, setWidth] = useState<number | undefined>();

  const handleEditClick = (): void => {
    setIsEditing(true);
  };

  const handleChange = (value: string): void => {
    if (value !== label) {
      setWidth(undefined);
      onChange?.(value);
    }
  };

  const handleOnBlur = (event: FocusEvent<HTMLInputElement>): void => {
    handleChange(event.currentTarget.value);
    setIsEditing(false);
  };

  const [lastValue, setLastValue] = useState<string>('');

  const handleKeyDown = (event: KeyboardEvent<HTMLInputElement>): void => {
    const value = event.currentTarget.value;
    if (lastValue.length < value.length) {
      setWidth(event.currentTarget.scrollWidth);
    }
    setLastValue(value);

    if (event.key === 'Enter') {
      handleChange(value);
      setIsEditing(false);
    }
  };

  return (
    <div className={c(styles.tab_editable, className)} style={{ width }}>
      <div className={c(styles.label)}>
        <span className={c(styles.text)}>{label || '...'}</span>
        {editable && (
          <button
            className={c(styles.edit_button)}
            onClick={handleEditClick}
            title="Edit tag"
          >
            <IconEdit className={c(styles.edit_icon)} />
            <i>Edit tag</i>
          </button>
        )}
      </div>
      {isEditing && (
        <input
          aria-label="Edit tag"
          autoFocus
          className={c(styles.input)}
          defaultValue={label}
          onBlur={handleOnBlur}
          onKeyDown={handleKeyDown}
          type="text"
        />
      )}
    </div>
  );
}
