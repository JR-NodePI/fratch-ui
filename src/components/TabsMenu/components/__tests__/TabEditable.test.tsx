import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, expect, it, vi } from 'vitest';

import TabEditable from '../TabEditable';
import { type TabEditableProps } from '../TabEditableProps';

describe('TabEditable', () => {
  const setup = (props: TabEditableProps = {}) =>
    render(<TabEditable {...props} />);

  it('should render properly', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('should render with label', () => {
    const { container } = setup({ label: 'label-test' });
    expect(container).toMatchSnapshot();
  });

  it('should render with className', () => {
    const { container } = setup({ className: 'className-test' });
    expect(container).toMatchSnapshot();
  });

  it('should render with editable', () => {
    const { container } = setup({ editable: true });
    expect(container).toMatchSnapshot();
  });

  it('should render in edition mode', async () => {
    const { container } = setup({ label: 'label-value-test', editable: true });

    await userEvent.click(
      screen.getByRole('button', {
        name: /Edit tag/i,
      })
    );

    expect(container).toMatchSnapshot();
  });

  it('should perform onChange on input field blur', async () => {
    const onChange = vi.fn();
    setup({
      label: 'current-label',
      editable: true,
      onChange,
    });

    await userEvent.click(
      screen.getByRole('button', {
        name: /Edit tag/i,
      })
    );

    const inputElement = screen.getByRole('textbox', { name: /Edit tag/i });

    expect(inputElement).toBeInTheDocument();
    await userEvent.click(inputElement);
    await userEvent.type(inputElement, '-and-new-value');
    fireEvent.blur(inputElement);
    expect(inputElement).not.toBeInTheDocument();

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith('current-label-and-new-value');
  });

  it('should perform onChange on keyDown Enter', async () => {
    const onChange = vi.fn();
    setup({
      label: 'current-label',
      editable: true,
      onChange,
    });

    await userEvent.click(
      screen.getByRole('button', {
        name: /Edit tag/i,
      })
    );

    const inputElement = screen.getByRole('textbox', { name: /Edit tag/i });

    expect(inputElement).toBeInTheDocument();
    await userEvent.click(inputElement);
    await userEvent.type(inputElement, '-and-new-value');
    await userEvent.keyboard('{enter}');
    expect(inputElement).not.toBeInTheDocument();

    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledWith('current-label-and-new-value');
  });
});
