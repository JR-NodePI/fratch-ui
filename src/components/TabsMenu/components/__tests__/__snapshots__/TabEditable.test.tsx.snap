// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TabEditable > should render in edition mode 1`] = `
<div>
  <div
    class="_tab_editable_732c04"
  >
    <div
      class="_label_732c04"
    >
      <span
        class="_text_732c04"
      >
        label-value-test
      </span>
      <button
        class="_edit_button_732c04"
        title="Edit tag"
      >
        <svg
          aria-hidden="true"
          class="_icon_2be8fc _primary_2be8fc _edit_icon_732c04"
          fill="none"
          viewBox="0 0 21 21"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.418 17.861 1 20l2.139-6.418m4.279 4.279 10.7-10.7a3.027 3.027 0 0 0-2.14-5.165c-.802 0-1.571.319-2.139.886l-10.7 10.7m4.279 4.279-4.279-4.279m2.139 2.14 7.844-7.844m-1.426-2.853 4.279 4.279"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <i>
          Edit tag
        </i>
      </button>
    </div>
    <input
      aria-label="Edit tag"
      class="_input_732c04"
      type="text"
      value="label-value-test"
    />
  </div>
</div>
`;

exports[`TabEditable > should render properly 1`] = `
<div>
  <div
    class="_tab_editable_732c04"
  >
    <div
      class="_label_732c04"
    >
      <span
        class="_text_732c04"
      >
        ...
      </span>
    </div>
  </div>
</div>
`;

exports[`TabEditable > should render with className 1`] = `
<div>
  <div
    class="_tab_editable_732c04 className-test"
  >
    <div
      class="_label_732c04"
    >
      <span
        class="_text_732c04"
      >
        ...
      </span>
    </div>
  </div>
</div>
`;

exports[`TabEditable > should render with editable 1`] = `
<div>
  <div
    class="_tab_editable_732c04"
  >
    <div
      class="_label_732c04"
    >
      <span
        class="_text_732c04"
      >
        ...
      </span>
      <button
        class="_edit_button_732c04"
        title="Edit tag"
      >
        <svg
          aria-hidden="true"
          class="_icon_2be8fc _primary_2be8fc _edit_icon_732c04"
          fill="none"
          viewBox="0 0 21 21"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.418 17.861 1 20l2.139-6.418m4.279 4.279 10.7-10.7a3.027 3.027 0 0 0-2.14-5.165c-.802 0-1.571.319-2.139.886l-10.7 10.7m4.279 4.279-4.279-4.279m2.139 2.14 7.844-7.844m-1.426-2.853 4.279 4.279"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <i>
          Edit tag
        </i>
      </button>
    </div>
  </div>
</div>
`;

exports[`TabEditable > should render with label 1`] = `
<div>
  <div
    class="_tab_editable_732c04"
  >
    <div
      class="_label_732c04"
    >
      <span
        class="_text_732c04"
      >
        label-test
      </span>
    </div>
  </div>
</div>
`;
